riptide-failsafe/src/test/java/org/zalando/riptide/failsafe/FailsafePluginCustomExecutorTest.java

```diff
@@ -0,0 +1,142 @@
+        AtomicInteger counter = new AtomicInteger();
+
+        public CountingExecutorService(int corePoolSize,
+                                       int maximumPoolSize,
+                                       long keepAliveTime,
+                                       TimeUnit unit,
+                                       BlockingQueue<Runnable> workQueue) {
+            super(corePoolSize, maximumPoolSize, keepAliveTime, unit, workQueue,
+                    Executors.defaultThreadFactory());
+        }
+
+        void resetCounter() {
+            counter.set(0);
+        }
+
+        @Override
+        public void execute(@NotNull Runnable command) {
+            super.execute(() -> {
+                log.info("Failsafe executor runnable");
+                counter.incrementAndGet();
+                command.run();
+            });
+        }
+    }
+
+    private final MockWebServer server = new MockWebServer();
+
+    private final CloseableHttpClient client = HttpClientBuilder.create().build();
+    private final ApacheClientHttpRequestFactory factory = new ApacheClientHttpRequestFactory(client);
+
+    private final CountingExecutorService countingExecutor = new CountingExecutorService(3, 3,
+            0L, TimeUnit.MILLISECONDS,
+            new LinkedBlockingQueue<>());
+
+    private final Http unit = Http.builder()
+            .executor(newFixedThreadPool(2))
+            .requestFactory(factory)
+            .baseUrl(getBaseUrl(server))
+            .converter(createJsonConverter())
+            .plugin(new FailsafePlugin()
+                    .withPolicy(Timeout.of(Duration.ofSeconds(1)))
+                    .withExecutor(countingExecutor)
+            )
+            .build();
+
+    private static MappingJackson2HttpMessageConverter createJsonConverter() {
+        final MappingJackson2HttpMessageConverter converter = new MappingJackson2HttpMessageConverter();
+        converter.setObjectMapper(createObjectMapper());
+        return converter;
+    }
+
+    private static ObjectMapper createObjectMapper() {
+        return new ObjectMapper().findAndRegisterModules()
+                .disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES);
+    }
+
+    @AfterEach
+    void tearDown() throws IOException {
+        client.close();
+        server.shutdown();
+        countingExecutor.resetCounter();
+    }
+
+    @Test
+    void shouldUseCustomExecutor() throws InterruptedException {
+        server.enqueue(emptyMockResponse());
+
+        int invocationCount = 5;
+        var futures = IntStream.range(0, invocationCount).mapToObj(i -> {
+            server.enqueue(emptyMockResponse());
+
+            return unit.get("/foo")
+                    .call(pass());
+        }).toList();
+
+        CompletableFuture.allOf(futures.toArray(new CompletableFuture[0]))
+                .join();
+
+        verify(server, invocationCount, "/foo");
+        assertEquals(invocationCount, countingExecutor.counter.get());
```

Comment:
I suggest to stick with one assertion library. In this test class, you mix Hamcrest and JUnit assertions.

Reply:
Fixed